src/v/cluster/shard_placement_table.cc

```diff
@@ -1221,11 +1221,15 @@ shard_placement_table::prepare_transfer(
         }
 
         if (is_initial) {
-            state._is_initial_for = std::nullopt;
-            if (state.is_empty()) {
-                _states.erase(ntp);
+            if (state._is_initial_for == expected_log_rev) {
+                state._is_initial_for = std::nullopt;
+                if (state.is_empty()) {
+                    _states.erase(ntp);
+                }
             }
+
             ret.is_finished = true;
+
             co_return ret;
```

Comment:
Prior to checking destination readiness, on lines 1113-1116, we perform the same check, but return with an error if it fails. And here we return with immediate success. Why, what's the difference?

Reply:
In the first case we haven't yet done anything and check if the transfer needs to be done. If there is inconsistency, we return an error and wait for the next reconciliation round. In the second case the transfer has already been done successfully, so we just need to clean up the local state on our shard.

In practice there is not much difference because in both cases we'll do another reconciliation round, but at least we'll signal to the caller if the transfer has been done.